service/history/historyEngine.go

```diff
@@ -690,108 +698,173 @@ func (e *historyEngineImpl) QueryWorkflow(
 	ctx ctx.Context,
 	request *h.QueryWorkflowRequest,
 ) (retResp *h.QueryWorkflowResponse, retErr error) {
-	domainCache, err := e.shard.GetDomainCache().GetDomainByID(request.GetDomainUUID())
+
+	context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetRequest().GetExecution())
 	if err != nil {
 		return nil, err
 	}
-	context, release, err := e.historyCache.getOrCreateWorkflowExecution(ctx, request.GetDomainUUID(), *request.GetExecution())
+	defer release(retErr)
```

Comment:
there is a difference between
`defer release(retErr)` vs `defer func() { release(retErr) } ()`
the first one basically get the value of retErr right away, while the second one will capture the retErr when defer function is invoked

here we should use the second one

Reply:
Thanks for catching this